Try: Improve permalink editing

earnjam opened the pull request:
## Description
Allows immediate editing of permalinks, without requiring a save. 
Fixes: #7129, #12714, #12031

Notes:
- In order edit the permalink before a save, we need to know what the permalink template structure is on page load. To get this, we need to modify `get_sample_permalink()` in core to add the `auto-save` status to the list of statuses that are faked to publish. See: [/wp-admin/includes/post.php#L1310](https://github.com/WordPress/WordPress/blob/47d32decd6668b3ccf100acc8a40aa0a9fc35003/wp-admin/includes/post.php#L1310). In the PR, I just used the `get_sample_permalink` filter at the end of the function and passed the parameters back through it, pretending it was published. I will need to open a separate core ticket to address this for whenever it's merged. 

## How has this been tested?
Extensive testing across various post statuses and different orders of editing the slug and/or title and the different methods of saving, publishing and transitioning the post status.

swissspidy commented on packages/editor/src/components/post-permalink/index.js:
```diff
@@ -78,25 +103,35 @@ class PostPermalink extends Component {
 				}
 
-				{ isEditable && ! isEditingPermalink &&
+				{ isEditable && ! isEditingPermalink && ! isNew &&
```
Can we have the same `! isNew` check for the `Change Permalinks` button? That shouldn't be displayed for new posts.

earnjam replied:
Now that I look at this again, this isn't actually doing anything, so I could drop this check. The entire component returns `null` on render if `isNew` is true ([See line 82](https://github.com/WordPress/gutenberg/pull/12009/files#diff-21ca0f66cf1102449de446bf6bf2fed2R82)), so it's always false at this point.

afercia commented:
Related: https://github.com/WordPress/gutenberg/issues/12031

earnjam commented:
Rebased and resolved merge conflicts. Still need to apply the updates to the sidebar panel too, but not sure the level of refactoring that should be done at this point right before an RC. 

Should note again that this PR would require a core patch at merge time.

youknowriad commented:
Thanks @earnjam for this PR and sorry for the delays, I'd like to see #7129 fixes if we can. It seems that this PR does a lot of things in parallel, do you think it's possible to split it into steps or does it only make sense as a single PR?

earnjam commented:
@youknowriad I'll look over this again today and see if I can split it up a bit. I think the only thing we could split out would be the application of the preview functionality, though I do think that would be nice to have. 

The other question around this, would it be more helpful to have a shared permalink component to use between both the title and the sidebar permalink editor instead of duplicating the effort?

youknowriad commented:
> The other question around this, would it be more helpful to have a shared permalink component to use between both the title and the sidebar permalink editor instead of duplicating the effort?

IMO, the title permalink should be deprecated at some point, but if it's easy to factorize a shared component, then, yes, it sounds like a great idea.

earnjam commented:
I rebased master, stripped out the preview functionality, moved the slug logic to a new `getEditedPostSlug` selector, and cleaned up the commits to make it easier to follow/read the changes here. 

I think this probably the minimum required to fix #7129 to allow permalink editing immediately without requiring any saves. 

There are more improvements that I'll probably try to make in separate PRs, such as recreating the logic from the `sanitize_permalink()` php function to make it a more exact match to what will be saved.

earnjam commented:
I probably should add some tests for this 🤔

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -2209,6 +2210,40 @@ export function getPermalink( state ) {
+	// then there may be a php generated slug available on the current_post.
+	// isSaving check is required because the title edit is cleared before the
+	// generated_slug value is set from the response.
+	const { postName } = getPermalinkParts( state );
```
There are failures in the E2E suite, which appear legitimate and may stem from this snippet of code.

```
● Sidebar Permalink Panel › should not render link panel when post is publicly queryable but not public
    expect(jest.fn()).not.toHaveErrored(expected)
    Expected mock function not to be called but it was called with:
    ["TypeError: Cannot read property 'postName' of null
```

What are the possible return values of `getPermalinkParts` ? Are we certain it's an object in such a way that we can confidently destructure it?

earnjam replied:
Yep, in this case non public or non publicly queryable posts return null on `getPermalinkParts()` because they don't have a `permalink_template` attribute. But I realized I could get what I need more directly than that since all I really need here is the `generated_slug` post attribute.

earnjam commented:
In writing some tests for this, so noticed that autosaves are no longer returning a `generated_slug` value. When I first worked on this 3+ months ago they were, so I used it to help calculate more accurate slugs. Seems related to removing the autosaves endpoint from the plugin and using the one from core instead. Not sure what specifically changed when integrating it, so I'll need to investigate and possibly rework the logic here a bit.

earnjam commented:
Looks like we were [manually adding by filtering the rest response](https://github.com/WordPress/gutenberg/blob/v4.9.0/lib/rest-api.php#L194-L224), but when that was migrated to core, it was [only applied to the posts controller](https://core.trac.wordpress.org/changeset/43720).

earnjam commented:
I opened two Trac tickets for the required core changes to make this work:
1. Modify `get_sample_permalink()`: https://core.trac.wordpress.org/ticket/46266
2. `generated_slug` on autosaves endpoint: https://core.trac.wordpress.org/ticket/46267

earnjam commented:
Rebased, updated and now working properly. I also added some unit tests for the new selector. 

The test that's failing is the same e2e test that everything on master is failing right now.

aduth commented:
This one is going to need some updates, in mind of #13716 (post effects refactored) and soon again with #7945 (autosave behaviors moved into `core` state). I feel bad for letting this one languish, so I'd like to help as I can. Do you think it might be good to wait for #7945 to land since it will likely have some significant impact? Would you have some feedback on the direction there, perhaps?

earnjam commented:
I agree on #7945 merging first, then updating this. I'd also love to see this land in WP 5.2, so don't want to let it sit too much longer. 

I'll take a look at 7945 and add any feedback if I have any.

SergeyBiryukov commented:
Related: #12907

talldan commented:
Hi @earnjam, is this something you plan to pick up again? I've added the `stale` label, as it seems this has a number of merge conflicts and hasn't been updated in a while.

earnjam commented:
@talldan Yeah, I'd like to get this in before 5.3. I'd already fixed the merge conflicts locally, just need to confirm everything is working and push. 

I'll either get it updated during contributor day at WC Boston this weekend, or early next week.

bartczyz commented:
Hi @earnjam, any updates on this one? Any way I can help?

mcsf commented:
This branch is fairly outdated. Is this PR to be revived here, or should we close it?

earnjam commented:
Just an update here...I've start refreshing this stale PR. Since it requires a PHP change on the WP side too, I'd like to get it ready for early in the WordPress 5.5 release cycle.

earnjam committed:
Filter REST API to return permalink_template

earnjam committed:
Add new getEditedPostSlug selector

earnjam committed:
Update auto-generated docs

earnjam committed:
Make permalink editing more timely/accurate in title editor

earnjam committed:
Make permalink editing more timely/accurate in sidebar panel

earnjam committed:
Apply minor code style changes

earnjam committed:
Update e2e tests to reflect that panel should always show

earnjam committed:
Add unit tests for getEditedPostSlug selector

earnjam commented:
I have rebased and updated the PR. 

I stripped out the portion of this that returned and used the `generated_slug` from an autosave because there are some complexities with regards to when `wp_unique_post_slug()` is actually run and some UX decisions to be made for whether the editor should reflect what is actually currently saved in the database, or what will happen after publish. I'll open another issue to discuss how we can improve that.

Also will open one for doing a better sanitization of the title that is closer to `sanitize_title()` from the PHP side.

earnjam commented:
The broken e2e test is because the permalink editor in above title shows up earlier than before and that test wasn't expecting it to be there. I'll get that fixed.

earnjam committed:
Don't show title permalink editor until a save

This somewhat diminishes the effects of this pull request, but it will
help it pass the e2e tests and also streamlines writing a bit. Reverts
back to preventing showing the permalink editor in the title until after
a manual or autosave. It does at least show the actual permalink on
the autosave and users can edit the slug at any time using the
document sidebar.

earnjam commented:
Fixed e2e tests by having the title permalink editor not show until the first save or autosave (same as it is currently). It _does_ however allow you to edit it after an autosave **and** shows the correct permalink structure, which is different from the current functionality.

You can also edit using the sidebar panel at any time, including before any save has occurred.

youknowriad reviewed (commented):
The code is looking very good here. Do you think we should land this before the the core patch https://core.trac.wordpress.org/ticket/46266 or maybe wait until the approach is validated. Who do you think can validate that patch?

earnjam commented:
@youknowriad I'm going to loop @desrosj in today to get his feedback on the core side and see if we should validate with anyone else, but I don't anticipate it being an issue.

youknowriad reviewed (approved):
This seems to work well for me. 👍

youknowriad commented:
Can we rebase and land this?

earnjam commented:
Rebased. I'll wait for tests to finish and then land it.

youknowriad commented:
Great work here and thanks for pushing it to the finish line.

youknowriad commented:
Should we still close the three issues mentioned here #7129, #12714, #12031

earnjam commented:
Yep, all 3 are closed now. The last one had some additional discussion on it, but the original basis of the ticket is resolved. I told Andrea he could open it back up if he felt it wasn't resolved.

paaljoachim commented:
I found it interesting that clicking Edit having the save button visible one could click it or just click outside to automatically close it. Perhaps this method can be used with other features as well. Instead of adding a cancel button one just clicks outside the button and the area it belongs to. 
I am thinking about the creation of reusable block and one wants to cancel the creation. https://github.com/WordPress/gutenberg/issues/12940
